src/browser/screen.js

```diff
@@ -67,86 +101,259 @@ function ScreenAdapter(options, screen_fill_buffer)
+
+        let txt_i, chr, chr_flags, chr_blinking, chr_font_offset, chr_bg_rgba, chr_fg_rgba;
+        let fg, bg, fg_r=0, fg_g=0, fg_b=0, bg_r=0, bg_g=0, bg_b=0;
+        let gfx_i, gfx_end_y, gfx_end_x, gfx_ic, glyph_i;
+        let draw_cursor, row, col, n_rows_rendered=0;
+
+        for(row = 0, txt_i = 0; row < text_mode_height; ++row)
+        {
+            if(!changed_rows[row])
+            {
+                txt_i += txt_row_size;
+                continue;
+            }
+
+            ++n_rows_rendered;
+            gfx_i = row * gfx_row_size;
+
+            for(col = 0; col < text_mode_width; ++col, txt_i += TEXT_BUF_COMPONENT_SIZE, gfx_i += gfx_col_step)
+            {
+                chr = text_mode_data[txt_i + CHARACTER_INDEX];
+                chr_flags = text_mode_data[txt_i + FLAGS_INDEX];
+                chr_blinking = chr_flags & FLAG_BLINKING;
+                chr_font_offset = chr_flags & FLAG_FONT_PAGE_B ? font_B_offset : font_A_offset;
+                chr_bg_rgba = text_mode_data[txt_i + BG_COLOR_INDEX];
+                chr_fg_rgba = text_mode_data[txt_i + FG_COLOR_INDEX];
+
+                if(bg !== chr_bg_rgba)
+                {
+                    bg = chr_bg_rgba;
+                    bg_r = bg >> 16;
+                    bg_g = (bg >> 8) & 0xff;
+                    bg_b = bg & 0xff;
+                }
+
+                if(chr_blinking && ! blink_visible)
+                {
+                    if(fg !== bg) {
+                        fg = bg;
+                        fg_r = bg_r;
+                        fg_g = bg_g;
+                        fg_b = bg_b;
+                    }
+                }
+                else if(fg !== chr_fg_rgba)
+                {
+                    fg = chr_fg_rgba;
+                    fg_r = fg >> 16;
+                    fg_g = (fg >> 8) & 0xff;
+                    fg_b = fg & 0xff;
+                }
+
+                draw_cursor = cursor_visible && cursor_gfx_i === gfx_i;
+
+                glyph_i = (chr_font_offset + chr) * font_size;
+
+                gfx_end_y = gfx_i + gfx_row_size;
+                for(; gfx_i < gfx_end_y; gfx_i += gfx_line_step)
+                {
+                    gfx_end_x = gfx_i + gfx_col_size;
+                    for(; gfx_i < gfx_end_x; gfx_i += 4)
+                    {
+                        if(font_bitmap[glyph_i++])
+                        {
+                            graphical_text_buffer[gfx_i]   = fg_r;
+                            graphical_text_buffer[gfx_i+1] = fg_g;
+                            graphical_text_buffer[gfx_i+2] = fg_b;
+                        }
+                        else
+                        {
+                            graphical_text_buffer[gfx_i]   = bg_r;
+                            graphical_text_buffer[gfx_i+1] = bg_g;
+                            graphical_text_buffer[gfx_i+2] = bg_b;
+                        }
+                    }
+                }
```

Comment:
That was the initial idea two or so months ago, but I soon came to realize that [`putImageData()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData) does not support changing the colors of the monochrome source bitmap, so this would unfortunately only work in monochrome mode.

I'm fully aware that `render_changed_rows()` implements a costly set of nested loops. There are two alternate ideas I have for this method in the back of my mind that would reduce costs: WebFonts or WASM.

* The first idea is to render VGA fonts dynamically into browser fonts, and then use classical text mode with these dynamically created web fonts. MIT-licensed library [opentype.js](https://opentype.js.org/) would seem a good starting point, and in their [issue 62](https://github.com/opentypejs/opentype.js/issues/62#issuecomment-2067820197) is a nice discussion about how to load such dynamically created fonts dynamically back into the browser. There are some downsides to this which kept me from pursuing this further:  
  * The discrete VGA glyph bitmaps would need to be vectorized to each glyph's outline, it's not enough to map pixels to squares, neighbouring squares need to be merged into larger shapes or it looks like a pixel cloud.
  * Browser's font scaling is not always precise, and
  * the recommended method to load the dynamically created font into the web page looks very costly, however font updates by the guest OS are usually extremely rare.
  
  Mabye the new [FontFace API](https://developer.mozilla.org/en-US/docs/Web/API/FontFace) could be useful for this, too.

* The second idea would be to rewrite this algorithm in WASM/rust. I'm still too new to the rust language and ecosystem to properly reason about this. I understand this would have to be asynchronous? It would be nice to write a synchronous WASM function where I could pass arguments by reference (to avoid copying these large buffers around). Writing the actual function would not be an issue for me, but I'd like to discuss this upfront cause I'm not sure how to go about it.

The second idea would be much more streamlined with the current code, the first would blow up everything again.

Any ideas and opinions would be very welcome here.

Reply:
> That was the initial idea two or so months ago, but I soon came to realize that [putImageData()](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData) does not support changing the colors of the monochrome source bitmap, so this would unfortunately only work in monochrome mode.

I see. There's still a possibility to do it using [drawImage](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) with certain [globalCompositeOperation](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Compositing#globalcompositeoperation) options, e.g.: https://stackoverflow.com/a/58077837

But I'd also merge it as it is now.

> Any ideas and opinions would be very welcome here.

Both of those suggestions sound a bit overkill for me.

Reply:
> There's still a possibility to do it using [drawImage](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) with certain [globalCompositeOperation](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Compositing#globalcompositeoperation) options, e.g.: https://stackoverflow.com/a/58077837

Interesting, I thought that wasn't possible because I had wrong ideas about globalcompositeoperation.

I might take a closer look some day but leave it as it is for now.

> Both of those suggestions sound a bit overkill for me.

Agreed :)

Reply:
I've fully reimplemented rendering using drawImage() and globalCompositeOperation, it works really great!

I'll commit this tomorrow.

Reply:
See commit fcb634b.

It passes all relevant tests with the current releases of Firefox and Chrome under Windows.

To make it a bit more obvious how this rendering method works I've put in a few comments.